Make .configure() configure network settings for both 1.x & 2.x

pimterry opened the pull request:
Connects to https://github.com/resin-io/resin-cli/issues/494
Connects to https://github.com/resin-io/resin-cli/issues/537

This PR does a lot of exciting things. Primarily, it ensures that resin-device-init knows how to initialize a device's network, not just its config.json. Lots of the logic to do that comes from the [reconfix test framework's schemas](https://github.com/resin-io/reconfix/tree/master/tests/integration/fixtures), and the [existing logic](https://github.com/resin-io/resin-cli/blob/master/lib/actions/local/configure.coffee) to do this for ResinOS 2.x images in the CLI (which will be removed shortly to start using this instead).

On top of that, this upgrades to resin-image-fs v4 (since that's a prerequisite), adds logic for OS version detection (some context [in flowdock](https://www.flowdock.com/app/rulemotion/r-resinos/threads/o2WC69hHrAv7fXVQ5gE4yp2xI2x)) since we need that to decide which type of network config to do, and updates the npm build script to always rebuild the docs, and build before you test, since that's easy to forget otherwise.

Some of this is arguably a little hacky, though I've tried to minimize that. A lot of that is due to limitations and outright bugs in the current reconfix implementation. I'm just working around those for now, since the JS reconfix is deprecated, and we should be replacing those parts with Rust reconfix soon-ish hopefully.

This isn't totally done - I still need to add some proper tests for this, but those are slightly tricky. It'd be useful if you could start on reviewing the implementation logic in the meantime though, and I'll patch some add some more thorough tests for this later too.

zvin commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+			DEFAULT_CONNECTION_FILE
+		)
+
+# Taken from https://goo.gl/kr1kCt
```
Can we use the real url in this comment?

pimterry replied:
This is taken from the CLI code, I suspect it's like this because the real URL is https://github.com/resin-os/meta-resin/blob/3b7c18a976ed882c6770886ba424ded243d59596/meta-resin-common/recipes-connectivity/networkmanager/files/resin-sample.ignore which doesn't exactly roll off the tongue, and is basically the fully width of my screen.

I'd rather keep the shortened version personally.

zvin commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+			],
+			template: {
+				'service_home_ethernet': {
+					'Type': 'ethernet',
```
NIT: no comma needed here.

zvin commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+		}
+		{
+			domain: [
+				[ 'network_config', 'service_home_ethernet' ],
```
NIT: no comma needed here.

zvin commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+			location:
+				getConfigPathDefinition(manifest, '/config.json', false)
+		network_config:
+			type: 'ini',
```
NIT: no comma needed here.

zvin commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+		network_config:
+			type: 'ini',
+			location:
+				parent: 'config_json',
```
NIT: no comma needed here.

zvin commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+	mapper: [
+		{
+			domain: [
+				[ 'config_json', 'wifiSsid' ],
```
NIT: no comma needed here.

zvin commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+	.then ->
+		schema = getOS2WifiConfigurationSchema(manifest, answers)
+		if manifest.configuration.config.image
+			image = path.join(image, manifest.configuration.config.image)
```
Anwswer to myself: that must be for edison images.

pimterry replied:
Exactly. If you try to configure a single image as an edison image you'll get errors complaining that `myimage.img/resin-image-edison.hddimg` does not exist, but imo that's not an unreasonable error, and it's a pretty rare case (and trying to configure a raspberry pi image to work for an edison is always going to blow up eventually).

zvin commented on lib/utils.coffee:
```diff
@@ -15,9 +15,11 @@ limitations under the License.
 rindle = Promise.promisifyAll(require('rindle'))
 path = require('path')
 stringToStream = require('string-to-stream')
+streamToString = require('stream-to-string')
```
Why use `streamToString` instead of `rindle.extractAsync` ?

pimterry replied:
Because I didn't know it existed :smiley:. Fixed.

pimterry commented on lib/init.coffee:
```diff
@@ -56,10 +59,27 @@ utils = require('./utils')
+				if not majorVersion? || majorVersion == 2
+					network.configureOS2Network(image, manifest, options)
+				if not majorVersion? || majorVersion == 1
+					network.configureOS1Network(image, manifest, options)
```
Yep, it'll do undefined things if you try to write changes in parallel. I've made it wait on these synchronously instead.

pimterry commented on lib/network.coffee:
```diff
@@ -0,0 +1,294 @@
+		contents.files ?= {}
+		contents.files['network/network.config'] ?= ''
+		imagefs.writeFile(configFilePath, JSON.stringify(contents))
+	.catch fileNotFoundError, ->
```
> I think you can use .catch { code: 'ENOENT' }, -> directly with Bluebird.
 
Sadly, this is using the promise returned from resin-image-fs, which uses an old version of bluebird that can't do this, and throws https://github.com/petkaantonov/bluebird/wiki/Error:-Catch-filter-must-inherit-from-Error-or-be-a-simple-predicate-function instead.

> If this catch is here to catch errors of imagefs.readFile(configFilePath), can we put it right after it (before JSON.parse)?

Done.

pimterry commented:
Updated to fix all the comments, except a couple I've replied to separately, and I've also added proper tests for network configuration.

The tests are failing on travis because our `ext2fs` module uses `const`, and doesn't actually support Node 4. We don't support Node 4 in the CLI either though, so we could just drop those tests drop support completely here too, unless there's some other reason that needs this to work in Node 4? We could probably make `ext2fs` backward compatible if we need to, but if not it'd be nice to keep it modern.

Meanwhile, the tests are failing on appveyor because ext2fs doesn't build on windows. This is going to be a bigger problem, since it really needs to build reliably on windows with nothing more than [`windows-build-tools`](https://www.npmjs.com/package/windows-build-tools), or none of our windows users are going to successfully install it. Right now, with that package installed, the current error is:

```
C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.Cpp.Platform.targets(64,5):
error MSB8020: The build tools for v140 (Platform Toolset = 'v140') cannot be found. To
build using the v140 build tools, please install v140 build tools. Alternatively, you may
upgrade to the current Visual Studio tools by selecting the Project menu or right-click
the solution, and then selecting "Upgrade Solution...".
[C:\projects\resin-device-init-xpd6l\node_modules\ext2fs\build\ext2fs.vcxproj]
```

I'm not immediately sure how to fix that - any ideas? I can dig into it more, but some pointers in the right direction first would be really useful.

pimterry commented:
Node-ext2fs issues seem to now be resolved. Next issue that's currently breaking the tests is https://github.com/resin-io/reconfix/pull/33: the reconfix JS implementation breaks with directories on windows. Quick fix, _hopefully_ this will start passing once that's sorted & released.

pimterry commented:
It's alive! Seems to all be working happily on Windows now. I _think_ this means this PR is finally all good and ready to go.

@zvin @abrodersen can you review please?

zvin commented on lib/utils.coffee:
```diff
@@ -39,16 +40,116 @@ exports.getManifestByDeviceType = (image, deviceType) ->
 		image: image
-		partition:
-			primary: 1
+		partition: 1
```
This won't work on a jetson-tx2 where boot is the 12th partition.
Shouldn't matter as there is a fallback.

zvin commented on lib/utils.coffee:
```diff
@@ -39,16 +40,116 @@ exports.getManifestByDeviceType = (image, deviceType) ->
+exports.getImageOsVersion = (image) ->
+	Promise.using imagefs.read(
+		image: image
+		partition: 2
```
This won't work on a jetson where resin-rootA is the 13th partition and uses the new host apps tree.

pimterry replied:
Yep, this doesn't work for quite a few similar cases too, so there's a `.catchReturn(null)` below, which should work as the fallback. Returning `null` from here results in us configuring the image as both OS1 _and_ OS2 for any images where we can't get a specific version easily. That's a little messy, but acceptable for now, and shouldn't break anything.

The plan in future is to eventually stop needing to look at the root partition at all this - instead we'll have an `os-release` file inside the boot partition itself, so we can use the device types to find it as we do with config.json/system-connections/etc. @agherzan is looking at that here: https://github.com/resin-os/meta-resin/issues/906.

zvin commented on tests/e2e.coffee:
```diff
@@ -204,10 +252,11 @@ wary.it 'should add a config.json to an intel edison',
+		Promise.using imagefs.read(
+			path: '/config.json'
+			image: path.join(images.edison, 'resin-image-edison.hddimg')
+		), extract
```
Extract could be replaced with `rindle.extractAsync`.

pimterry replied:
True, good idea, but I'm going to ignore this for now anyway. I want to do a bunch of improvements to the tests (e.g. #26) anyway, and we use this `extract` method everywhere, so I'll probably replace it all with rindle as part of that.
